### PR TITLE
Execution pool implementation

### PR DESCRIPTION
**Pull request by Toru8P**

### CODE REVIEW

File: src/main/java/com/exactpro/th2/sim/impl/Simulator.java [hunk 1]
```diff
@@ -248,49 +287,103 @@ public void touchRule(@NotNull TouchRequest request, StreamObserver<Empty> respo
...
+            try {
+                List<MessageGroup> messageGroups = batch.getGroupsList();
+                int messageGroupsSize = messageGroups.size();
+                for (int messageGroupIndex = 0; messageGroupIndex < messageGroupsSize; messageGroupIndex++) {
+
+                    List<AnyMessage> messages = messageGroups.get(messageGroupIndex).getMessagesList();
+                    int messageGroupSize = messages.size();
+                    for (int messageIndex = 0; messageIndex < messageGroupSize; messageIndex++) {
+
+                        AnyMessage message = messages.get(messageIndex);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/sim/impl/Simulator.java:301]:
```suggestion
                for (MessageGroup messageGroup : batch.getGroupsList()) {
                    for (AnyMessage message : messageGroup.getMessagesList()) {
```

File: src/main/java/com/exactpro/th2/sim/impl/Simulator.java [hunk 2]
```diff
@@ -248,49 +287,103 @@ public void touchRule(@NotNull TouchRequest request, StreamObserver<Empty> respo
...
+                        if (message.hasMessage()) {
+                            handleMessage(message.getMessage(), relation);
+                        } else {
+                            logger.warn("Unsupported format of incoming message: {}", message.getKindCase().name());
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                if (logger.isErrorEnabled()) {
+                    logger.error("Can`t handle batch = {}", TextFormat.shortDebugString(batch), e);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/sim/impl/Simulator.java:311]: > TextFormat

Use JSON instead

File: src/main/java/com/exactpro/th2/sim/impl/Simulator.java [hunk 3]
```diff
@@ -248,49 +287,103 @@ public void touchRule(@NotNull TouchRequest request, StreamObserver<Empty> respo
...
             return;
         }
 
-        List<SimulatorRuleInfo> triggeredRules = relationRules.stream()
-                .map(ruleIds::get)
-                .filter(ruleInfo -> ruleInfo.checkAlias(message) && ruleInfo.getRule().checkTriggered(message))
-                .collect(Collectors.toList());
+        int triggeredSize = triggeredRules.size();
+        for (int i = 0; i < triggeredSize; i++) {
+            SimulatorRuleInfo triggeredRule = triggeredRules.get(i);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/sim/impl/Simulator.java:331]:
```suggestion
        for (SimulatorRuleInfo triggeredRule : triggeredRules) {
```

File: src/main/java/com/exactpro/th2/sim/impl/Simulator.java [hunk 4]
```diff
@@ -248,49 +287,103 @@ public void touchRule(@NotNull TouchRequest request, StreamObserver<Empty> respo
...
-                logger.trace("Only non default rules will process message ({}) due strategy: {}", msgType, strategy.name());
-                triggeredRules = nonDefaultRules;
+        var messageType = message.getMetadata().getMessageType();
+        if(relationRules == null) {
+            logger.trace("No related rules was found for message: {}", messageType);
+            return Collections.emptyList();
+        }
+
+        if (logger.isTraceEnabled()) {
+            logger.trace("Message {} checking against related rules: {}", messageType, relationRules.stream().map(integer -> ruleIds.get(integer).getRule().getClass().getSimpleName()).collect(Collectors.joining(", ")));
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/sim/impl/Simulator.java:353]:
```suggestion
            String rules = relationRules.stream()
                    .map(id -> ruleIds.get(id).getRule().getClass().getSimpleName())
                    .collect(Collectors.joining(", "));
            
            logger.trace("Message {} checking against related rules: {}", messageType, rules);
```

File: src/main/java/com/exactpro/th2/sim/impl/Simulator.java [hunk 5]
```diff
@@ -248,49 +287,103 @@ public void touchRule(@NotNull TouchRequest request, StreamObserver<Empty> respo
...
+        }
+
+        List<SimulatorRuleInfo> triggeredRules = new ArrayList<>();
+        int relationSize = relationRules.size();
+        for (int i = 0; i < relationSize; i++) {
+            SimulatorRuleInfo rule = ruleIds.get(relationRules.get(i));
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/sim/impl/Simulator.java:359]:
```suggestion
        for (Integer relationRule : relationRules) {
            SimulatorRuleInfo rule = ruleIds.get(relationRule);
```

File: src/main/java/com/exactpro/th2/sim/impl/SimulatorRuleInfo.java [hunk 6]
```diff
@@ -328,30 +268,20 @@ private AnyMessage prepareMessage(@NotNull AnyMessage msg) {
...
-                .getSessionAlias()
-                .equals(configuration.getSessionAlias());
+        String alias = configuration.getSessionAlias();
+        return alias == null || alias.isEmpty() || MessageUtils.getSessionAlias(message).equals(alias);
     }
 
-    private long checkDelay(long delay) {
-        if (delay < 0) {
-            throw new IllegalStateException("Negative delay in rule " + id + ": " + delay);
+    private void requirePositiveOrZero(long value, Supplier<String> messageSupplier) {
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/sim/impl/SimulatorRuleInfo.java:275]:
```suggestion
    private void requireNonNegative(long value, Supplier<String> messageSupplier) {
```

File: src/main/kotlin/com/exactpro/th2/sim/util/MessageBatcher.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    private val maxFlushTime: Long = 1000,
+    private val executor: ScheduledExecutorService,
+    private val onBatch: (MessageGroupBatch, String) -> Unit
+) : AutoCloseable {
+    private val batches = ConcurrentHashMap<String, MessageBatch>()
+
+    fun onMessage(message: RawMessage, relation: String) = batches.getOrPut(relation) { MessageBatch(relation) }.add(message.toGroup())
+    fun onMessage(message: Message, relation: String) = batches.getOrPut(relation) { MessageBatch(relation) }.add(message.toGroup())
+    fun onMessage(message: AnyMessage, relation: String) = batches.getOrPut(relation) { MessageBatch(relation) }.add(MessageGroup.newBuilder().addMessages(message).build())
+    fun onGroup(group: MessageGroup, relation: String) = batches.getOrPut(relation) { MessageBatch(relation) }.add(group)
```
Comment (ivandruzhinin) [inline src/main/kotlin/com/exactpro/th2/sim/util/MessageBatcher.kt:28]: Why do we batch by relation?

Comment (Topru333) [reply src/main/kotlin/com/exactpro/th2/sim/util/MessageBatcher.kt:28]: We can have pins with only one relation, grouping by relation and sending as group with attribute seemed like a good idea to me

File: src/main/java/com/exactpro/th2/sim/configuration/RuleConfiguration.java [hunk 8]
```diff
@@ -28,6 +28,9 @@ public String getSessionAlias() {
     }
 
     public void setSessionAlias(String sessionAlias) {
+        if (sessionAlias !=null && sessionAlias.isBlank()) {
```
[the remark this answers is not in the document]
Comment (Topru333) [reply src/main/java/com/exactpro/th2/sim/configuration/RuleConfiguration.java:31]: redunant check i think

File: src/main/java/com/exactpro/th2/sim/impl/Simulator.java [hunk 9]
```diff
@@ -186,10 +187,8 @@ public RuleID addRule(@NotNull IRule rule, @NotNull RuleConfiguration configurat
         String relation = configuration.getRelation();
 
         ruleIds.put(id, ruleInfo);
-        List<Integer> relationRules = relationToRuleId.computeIfAbsent(relation, (key) -> new CopyOnWriteArrayList<>());
-        if (!relationRules.contains(ruleInfo.getId())) {
-            relationRules.add(ruleInfo.getId());
-        }
+        Set<Integer> relationRules = relationToRuleId.computeIfAbsent(relation, (key) -> new CopyOnWriteArraySet<>());
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/sim/impl/Simulator.java:190]:
```suggestion
        Set<Integer> relationRules = relationToRuleId.computeIfAbsent(relation, (key) -> ConcurrentHashMap.newKeySet());
```

**Pull request merged**